text/0000-expand-math.md

```diff
@@ -0,0 +1,230 @@
+- Feature Name: Introduce new numerics -- `Rational`, `BigInt`, `BigFloat`, `Complex`
+- Start Date: 2022-02-28
+- RFC PR:
+- Pony Issue:
+
+# Summary
+
+This RFC proposes the introduction of new numeric types; in particular the addition of a type representing a fractional number (`Rational`), arbitrary precision integer (`BigInt`), arbitrary precision float (`BigFloat`), and complex number (`Complex`).
+
+# Motivation
+
+The primary motivation for adding these types to the stdlib is to have a single canonical implementation of them which allow interoperability of numeric types across the Pony ecosystem.
+
+# Detailed design
+
+I propose we add the aforementioned numeric types into `builtin` so they exist alongside the other standard numeric types. These introduced numeric types **must** existing within the current numeric type hierarchy by being compliant with existing numeric traits.
+
+## Numeric Hierarchy
+
+Current the Pony numeric type hierarchy is as follows:
+
+```mermaid
+classDiagram
+class Any
+<<interface>> Any
+class Real
+<<trait>> Real
+class FloatingPoint
+<<trait>> FloatingPoint
+class Integer
+<<trait>> Integer
+class SignedInteger
+<<trait>> SignedInteger
+class UnsignedInteger
+<<trait>> UnsignedInteger
+
+Any <-- Real
+Real <-- FloatingPoint
+Real <-- Integer
+Integer <-- SignedInteger
+Integer <-- UnsignedInteger
+
+FloatingPoint <-- F32
+FloatingPoint <-- F64
+
+UnsignedInteger <-- U8
+UnsignedInteger <-- U16
+UnsignedInteger <-- U32
+UnsignedInteger <-- U64
+UnsignedInteger <-- U128
+UnsignedInteger <-- USize
+UnsignedInteger <-- ULong
+
+SignedInteger <-- I8
+SignedInteger <-- I16
+SignedInteger <-- I32
+SignedInteger <-- I64
+SignedInteger <-- I128
+SignedInteger <-- ISize
+SignedInteger <-- ILong
+```
+
+This RFC introduces four more numeric types: `Rational`, `BigInt`, `BigFloat`, and `Complex`. These fit into the numeric type hierarchy in the following manner:
+
+```mermaid
+classDiagram
+class Any
+<<interface>> Any
+class Real
+<<trait>> Real
+class FloatingPoint
+<<trait>> FloatingPoint
+class Integer
+<<trait>> Integer
+class SignedInteger
+<<trait>> SignedInteger
+class UnsignedInteger
+<<trait>> UnsignedInteger
+
+Any <-- Real
+Any <-- Complex~Real~
+Real <-- FloatingPoint
+Real <-- Integer
+Real <-- Rational~Integer~
+Integer <-- SignedInteger
+Integer <-- UnsignedInteger
+Integer <-- BigInt
+
+FloatingPoint <-- F32
+FloatingPoint <-- F64
+FloatingPoint <-- BigFloat
+
+UnsignedInteger <-- U8
+UnsignedInteger <-- U16
+UnsignedInteger <-- U32
+UnsignedInteger <-- U64
+UnsignedInteger <-- U128
+UnsignedInteger <-- USize
+UnsignedInteger <-- ULong
+
+SignedInteger <-- I8
+SignedInteger <-- I16
+SignedInteger <-- I32
+SignedInteger <-- I64
+SignedInteger <-- I128
+SignedInteger <-- ISize
+SignedInteger <-- ILong
+```
+
+## Methods of Concern
+
+```pony
+trait val Real[A: Real[A] val] is
+  (Stringable & _ArithmeticConvertible & Comparable[A])
+  ...
+  new val min_value()
+  new val max_value()
+  ...
+```
```

Comment:
I understand that this has already been present, but given this re-design I'm not clear on the benefit of these constructors, but there's some obvious costs. Is there a use case for max_value() / min_value() that makes sense for all numbers? Given that this doesn't work for BigInt/BigFloat/similar types I'd be hesitant to place it at the top of the hierarchy.

Can we instead make a Bounded interface?

Reply:
I don't understand what you are suggesting. What's a "bounded interface"?

Reply:
I think he means a new interface whose name is `Bounded`, which has the min and max value methods on it.

Reply:
@jasoncarr0 is that what you meant? re: `Bounded` interface? if yes, did you specifically mean structural typing or were you using "interface" more loosely?

Reply:
I meant a literal Pony interface, that is:
These constructors would be part of an interface named `Bounded`, while the other methods would be part of this trait.

Reply:
Actually there is also an argument for removing Ordered, but only because this trait is really the lowest possible. Mostly that the implementations for things like matrices would be sketch but making them numerics makes sense and is powerful.